### PR TITLE
Retrieve the block results that include the event with the exact attributes in query

### PR DESCRIPTION
**Pull request by romac**
Closes: #2867 
Closes: #2868

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [x] Reviewed `Files changed` in the GitHub PR explorer.
- [x] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/relayer/src/chain/cosmos/query/tx.rs [hunk 1]
```diff
@@ -183,6 +185,27 @@ pub async fn query_packets_from_block(
...
+        &mut block_results
+            .end_block_events
+            .unwrap_or_default()
+            .into_iter()
+            .filter_map(|ev| filter_matching_event(ev, request, &request.sequences))
+            .map(|ev| IbcEventWithHeight::new(ev, height))
+            .collect(),
+    );
+    result.append(&mut begin_block_events);
+    result.append(&mut end_block_events);
```
Comment (ancazamfir) [inline crates/relayer/src/chain/cosmos/query/tx.rs:208]: The order should be: begin_block_events, result, end_block_events.
Maybe also rename `result` to `tx_results`.

File: crates/relayer/src/chain/cosmos.rs [hunk 2]
```diff
@@ -577,22 +578,25 @@ impl CosmosSdkChain {
...
-                    Order::Ascending,
+                    // There should only be a single match for this query, but due to
+                    // the fact that the indexer treat the query as a disjunction over
+                    // all events in a block rather than a conjunction over a single event,
+                    // we may end up with partial matches and therefore have to account for
+                    // that by fetching multiple results and filter it down after the fact.
+                    10,
+                    // Order them in descending order so we get the most recent blocks,
+                    // which are likely to be the most relevant.
+                    Order::Descending,
```
Comment (ancazamfir) [inline crates/relayer/src/chain/cosmos.rs:595]: I am not sure `Descending` has any advantage. We can keep it that way but change the comment or clarify why.

File: crates/relayer/src/chain/cosmos.rs [hunk 3]
```diff
@@ -577,22 +578,25 @@ impl CosmosSdkChain {
...
+                    // We only need the first page
                     1,
-                    1, // there should only be a single match for this query
-                    Order::Ascending,
+                    // There should only be a single match for this query, but due to
+                    // the fact that the indexer treat the query as a disjunction over
+                    // all events in a block rather than a conjunction over a single event,
+                    // we may end up with partial matches and therefore have to account for
+                    // that by fetching multiple results and filter it down after the fact.
+                    10,
```
Comment (ancazamfir) [inline crates/relayer/src/chain/cosmos.rs:592]: I think in the worst case we get N blocks where N is the number of channels. Maybe clarify this.
Also should we have a `const` for this?

**Comment by ancazamfir** [comment]
I think also fixes #2868

File: crates/relayer/src/chain/cosmos/query/tx.rs (hunk 1 above)
Comment (romac) [reply crates/relayer/src/chain/cosmos/query/tx.rs:208]: 8a59a565

File: crates/relayer/src/chain/cosmos/query/tx.rs [hunk 4]
```diff
@@ -183,7 +183,31 @@ pub async fn query_packets_from_block(
...
+            .filter_map(|ev| filter_matching_event(ev, request, &request.sequences))
+            .map(|ev| IbcEventWithHeight::new(ev, height))
+            .collect(),
+    );
+
+    let mut events = begin_block_events;
+    events.append(&mut tx_events);
+    events.append(&mut end_block_events);
+
+    Ok(events)
```
Comment (adizere) [inline crates/relayer/src/chain/cosmos/query/tx.rs:210]: Implementation note: Do I understood correctly that this changes here is strictly for #2868 ?

File: crates/relayer/src/chain/cosmos.rs [hunk 5]
```diff
@@ -577,22 +578,25 @@ impl CosmosSdkChain {
...
                 ))
                 .map_err(|e| Error::rpc(self.config.rpc_addr.clone(), e))?;
 
-            assert!(
-                response.blocks.len() <= 1,
-                "block_search: unexpected number of blocks"
-            );
-
-            if let Some(block) = response.blocks.first().map(|first| &first.block) {
+            for block in response.blocks.into_iter().map(|response| response.block) {
```
Comment (adizere) [inline crates/relayer/src/chain/cosmos.rs:599]: If I understand correctly, the changes in this file are particularly for fixing issue #2867 ?

Comment (ancazamfir) [reply crates/relayer/src/chain/cosmos.rs:599]: yes

**Comment by adizere** [comment]
What do you think about untangling this PR to separate the changes related to issue 2868 apart from issue 2867?

Context:
- I think we can merge the changes related to https://github.com/informalsystems/hermes/issues/2868 ASAP. They look solid.
- As for https://github.com/informalsystems/hermes/issues/2867 this is more complex.
    - Jasmina proposed a Tendermint-level fix in this PR https://github.com/tendermint/tendermint/pull/9759
    - With these changes in place we can query from Hermes using the new [`match.events`](https://github.com/tendermint/tendermint/pull/9759/files#diff-1279677bb1be6aca80473a8ff85bb14d9696926041e9c3756de24d7c12594661R236) flag.
    - This requires tendermint-rs changes, so it will be a bit more work to push that through, but it's a long term fix.
    - In order to unblock the Tendermint team, it would be good if we give them feedback on their PR.


Summary:
- We can fix #2868 by trimming down the present PR (or a new PR with the right changes)
- We can open a new PR to test E2E if the changes in Tendermint (https://github.com/tendermint/tendermint/pull/9759) are sufficient to fix #2867.
    - If yes, we give the feedback to TM team and go with that
    - If no, we retain the solution in this PR

**Comment by ancazamfir** [comment]
I'll take a look at the tendermint PR and run a custom hermes version that uses that. But it seems that we still need to temp fix hermes with the PR here until all chains move to the new tendermint version.

**Review by ancazamfir — APPROVED** [review]
lgtm, thanks @romac!

**Pull request merged**